burn-ndarray/src/ops/conv.rs

```diff
@@ -25,59 +25,119 @@ pub(crate) fn conv2d<E: FloatNdArrayElement>(
+                                        dilation_width,
+                                        dilation_height,
+                                    )
+                                {
+                                    for oh in 0..out_height {
+                                        // Construct a sub-slice view of the input row.
+                                        // This is done upfront so that rustc does not have to emit bounds checks
+                                        // in the hot loop below.
+                                        let ir = x.row(oh + kh).to_slice().unwrap();
+
+                                        let mut or = output.row_mut(oh);
+                                        let or = &mut or.as_slice_mut().unwrap()[0..out_width];
+
+                                        #[allow(clippy::needless_range_loop)]
+                                        for ow in 0..out_width {
+                                            let iw = ow + kw;
+                                            or[ow] += ir[iw] * k;
+                                        }
+                                    }
+                                } else {
+                                    for oh in 0..out_height {
+                                        // Construct a sub-slice view of the input row.
+                                        // This is done upfront so that rustc does not have to emit bounds checks
+                                        // in the hot loop below.
+                                        let ir = x
+                                            .row(oh * stride_height + kh * dilation_height)
+                                            .to_slice()
+                                            .unwrap();
+
+                                        let mut or = output.row_mut(oh);
+                                        let or = &mut or.as_slice_mut().unwrap()[0..out_width];
+
+                                        #[allow(clippy::needless_range_loop)]
+                                        for ow in 0..out_width {
+                                            let iw = (ow * stride_width) + (kw * dilation_width);
+                                            or[ow] += ir[iw] * k;
+                                        }
+                                    }
+                                }
                             }
                         }
                     }
-                }
-            }
 
-            if let Some(bias) = &bias {
-                for oh in 0..out_height {
-                    for ow in 0..out_width {
-                        output[[b, oc, oh, ow]] += bias.array[oc];
+                    if let Some(bias) = &bias {
+                        let bias = bias.array[oc];
+
+                        for oh in 0..out_height {
+                            let mut or = output.row_mut(oh);
+                            let or = &mut or.as_slice_mut().unwrap()[0..out_width];
```

Comment:
The code that extract the output row and the input row are duplicated at multiple places, and are quite hard to read. I would extract them into function, maybe with `#[inline(always)]`:

```rust
#[inline(always)]
fn output_row<'a>(output: &'a mut Array3<E>, oh: usize) -> &'a mut [E] {
    let mut or = output.row_mut(oh);
    &mut or.as_slice_mut().unwrap()[0..out_width]
}
```

We could do the same for input row. What are your thoughts?

Reply:
Unfortunately, I had to decompose these statements into two because `row_mut` creates an object that `as_slice_mut` borrows from.
In other words, such a function would have to return the `or` variable _and_ the slice created in order to make the borrow checker happy.

I agree that it's a bit dense - but I don't think a helper function would work in this case. The best I can do is add comments explaining everything that is happening here :)